Add a tutorial on creating an inset plot

sashansamarajeewa opened the pull request:
# Description

Fixes #3292 

This PR adds a tutorial on creating an inset plot. The related issue #3292 discusses 2 approaches to address this but I opted for the initial suggestion described in https://github.com/MakieOrg/Makie.jl/issues/3292#issuecomment-2483439053.

The tutorial consists of 3 main parts.

1. Explain what is an inset plot.
2. How to make an inset plot with an example.
3. Explain key elements and related information.

Output image

![image](https://github.com/user-attachments/assets/4c076c83-7589-4947-bc9c-9e70f423fc3e)


## Type of change

Delete options that do not apply:

- [x] Improving the Documentation

## Checklist

- [x] Added or changed relevant sections in the documentation

sashansamarajeewa commented:
@ffreyer @jkrumbiegel It would be great if someone can review this PR and provide feedback :)

ffreyer commented on docs/src/tutorials/inset-plot-tutorial.md:
````diff
@@ -0,0 +1,184 @@
+Indicate the zoomed section of the main plot and guide the eye to the inset plot using lines.
+
+```@figure inset
+zoom_lines!(ax_main, ax_inset)
+save("output.png", fig) # hide
````
I guess this is more personal preference but I'd like this to be `MakieExtras.zoom_lines!(...)` so you can easily see what functionality you need MakieExtras for. Maybe also move the `using MakieExtras` here to further clarify.

I think the `save` here is unnecessary and should be deleted

sashansamarajeewa replied:
Isn't `save` required to display the output image in the beginning with `![](output.png)`?

ffreyer replied:
Oh right, didn't see that

ffreyer commented on docs/src/tutorials/inset-plot-tutorial.md:
```diff
@@ -0,0 +1,184 @@
+
+### 2. `translate!` Function and Z-Ordering
+
+**z-order** (depth) determines the rendering order of elements, with higher z-values appearing in front of lower ones. This is critical for ensuring that the inset plot is visible above the main plot. By explicitly setting the z-value via translate! function, you can layer elements as needed. If translate! is omitted or the z-value is too low, the inset plot may render behind the main plot, makin ...
```
I think it would be good to give a bit background information here. If you search for "translate" in https://github.com/MakieOrg/Makie.jl/blob/master/src/makielayout/blocks/axis.jl and https://github.com/MakieOrg/Makie.jl/blob/master/src/makielayout/lineaxis.jl you'll find translates between -100 (Axis background) and +20 (frame lines (?)) / +10 (tick lines) / +0 (tick text, titles). User plots are naturally at 0. 

So you need at least +100 for the background of the inset to compete with (un-translated) user plots. Depending on when the plots are created you might need to be over 100.

If you would want the inset axis to draw partially outside the axis you'd need at least +120. (I'm not sure if layouting would let you do this though.) The maximum z  you're allowed to have is 10_000.

The tutorial should simplify that information. Maybe just mention that Axis spans a z range of -100 to +20 and adjust the translate you used before to something like 150 to be safe. And also mention that if you z-translate plots in the big Axis you may need to shift the inset Axis more. (I.e. you want `plot_translate < inset_axis_translate - 100`)

sashansamarajeewa replied:
Updated the content with more information.

jkrumbiegel commented:
I think this tutorial would be better suited for the MakieExtras docs as it describes how to use its functions, and not how to do a generic inset with Makie. I don't want to document other packages here if I don't have to as it increases the overall burden of keeping things up to date, plus, MakieExtras is specifically a package doing nonstandard things with some internals use etc.

SimonDanisch commented on docs/src/tutorials/inset-plot-tutorial.md:
````diff
@@ -0,0 +1,184 @@
+Start by adding CairoMakie backend package and MakieExtra package into the environment.
+
+```@example inset
+using Pkg
+Pkg.add("CairoMakie")
+Pkg.add("MakieExtra")
````
For what is MakieExtra even added? On first sight it all looks like normal Makie code.

sashansamarajeewa replied:
MakieExtra package is used to add [zoom_lines](https://juliaaplavin.github.io/MakieExtraDocs.jl/notebooks/examples.html#:~:text=end-,Zoom%20lines,-Whenever%20one%20axes) functionality. It marks which part of the main plot is shown in the inset axis.

sashansamarajeewa commented:
> I think the part with the zoom lines would be better suited for the MakieExtras docs as it describes how to use its functions, and not how to do generic operations with Makies objects. I don't want to document other packages here if I don't have to as it increases the overall burden of keeping things up to date, plus, MakieExtras is specifically a package doing nonstandard things with some internals use etc.

Do you suggest removing the MakieExtra package and zoom_lines function from the plot and keeping the rest?

jkrumbiegel commented:
Yeah probably, it's already hard to keep the docs in sync with our own package, much less if we add others as well. I'd be open to showing how to do your own zoomlines, although it's maybe a little involved for a tutorial.

SimonDanisch commented:
I think we should either port zoomlines to Makie or document it in Makie Extras and link it in the docs.
`projection` is considered API, albeit not a thought through and well tested one :( 
CC: @aplavin

ffreyer commented:
I think project is too much of a mess to advertise it. 

`space = :relative` is also not enough here because the area you're targeting is probably in data space.

sashansamarajeewa commented:
Does it make sense to point to the zoomlines [documentation](https://juliaaplavin.github.io/MakieExtraDocs.jl/notebooks/examples.html#:~:text=end-,Zoom%20lines,-Whenever%20one%20axes) within this tutorial or completely remove MakieExtra package and zoom_lines function from this tutorial?

sashansamarajeewa commented on docs/src/tutorials/inset-plot-tutorial.md:
````diff
@@ -0,0 +1,184 @@
+stock_price_inset = stock_price[50:70]
+line_inset = lines!(ax_inset, time_inset, stock_price_inset, color=:red)
+fig
+```
````
Added xlims and ylims for selecting the data ranges.

sashansamarajeewa commented:
Removed `MakieExtra` package and `zoom_lines!` function from this tutorial. Instead, added information on how to create a border to mark the zoomed section.

SimonDanisch commented:
Awesome :) We could maybe still link the docs for zoom_lines?

sashansamarajeewa commented:
> Awesome :) We could maybe still link the docs for zoom_lines?

Added information about `zoom_lines` function and linked the documkentation.

sashansamarajeewa commented:
Is it possible to build the artifacts if there are no more changes to be done?

sashansamarajeewa commented:
Build was failing due to missing a CHANGELOG entry. Updated CHANGELOG.md file.

ffreyer reviewed (approved):
I think this is good to merge now.  (Assuming the docs build doesn't fail)

> Build was failing due to missing a CHANGELOG entry. Updated CHANGELOG.md file.

We usually don't add docs changes there but maybe it's good to have this one. May help make people aware of it.

sashansamarajeewa commented:
Build has been successful and I hope this will be merged 🤞

SimonDanisch commented:
Great work :) I've checked the uploaded artifact of the documentation and it seems to all work well!

sashansamarajeewa commented:
Latest CHANGELOG.md changes was conflicting with this branch so merged it.
Let me know if there is anything else to be done before merging into master.
